### ISSUE TITLE
Unable to add large margins on page

### ISSUE THREAD

**Issue by JElgar** [espresso3389/pdfrx#239]
I am trying to add a large margin around the pdf page so the user can scroll and drag the pdf to the edge of the screen e.g. (in this screenshot the whole screen is the viewer but the pdf has been dragged right to the edge)
<img width="742" alt="image" src="https://github.com/user-attachments/assets/6447d1b7-baa8-452e-9d66-ea47c9b7e947">

I tried doing this by adding a large margin to the page but this caused the assertion error that checks `minWidth > 0` when rendering the interactive viewer. It look like this code does some logic to work out what the min width should be and if any dimension of the view size is greater than the half the margin then a negative number is returned

https://github.com/espresso3389/pdfrx/blob/6e52b2467f5e7e1fdedfca101b6eeff038ae3bd4/lib/src/widgets/pdf_viewer.dart#L674

You should be able to reproduce this issue by adding a large margin to a pdf (one larger than the view's smaller dimension).

Is this behaviour expected? If so is there an alternative way of achieving what Im looking for? Thank you!

**Comment by JElgar** [comment]
So I did a bit more poking on this and it seems I can get what I need if I override `_makeMatrixInSafeRange` in `_PdfViewerTransformationController`. The behavior I need is along the lines of this (which keeps 20% of the page on the screen at a minimum). 

```
  Matrix4 _makeMatrixInSafeRange(Matrix4 newValue) {
    _updateViewSizeAndCoverScale(_viewSize!);

    const minScale = 1.0;
    const minFractionOnScreen = 0.2;

    final position = newValue.calcPosition(_viewSize!);

    final newZoom = max(newValue.zoom, minScale);

    final hw = _viewSize!.width / 2 / newZoom;
    final hh = _viewSize!.height / 2 / newZoom;

    final minX = -hw + _layout!.documentSize.width * minFractionOnScreen;
    final maxX = hw + _layout!.documentSize.width * (1 - minFractionOnScreen);

    final minY= -hh + _layout!.documentSize.height * minFractionOnScreen;
    final maxY = hh + _layout!.documentSize.height * (1 - minFractionOnScreen);

    final x = position.dx.clamp(minX, maxX);
    final y = position.dy.clamp(minY, maxY);

    return _calcMatrixFor(Offset(x, y), zoom: newZoom, viewSize: _viewSize!);
  }
```

Im not sure how this can be applied more generally though? Do you think it would make sense to provide a mechanism to override the transformer or is there a better way?

**Comment by espresso3389** [comment]
With 1.0.86, you can use PdfViewerParams.normalizeMatrix to customize the transform matrix restriction; customizing existing logic on _PdfViewerState._makeMatrixInSafeRange.

**Comment by JElgar** [comment]
Amazing thank you!

**Comment by JElgar** [comment]
Ah but quick question on that. Is there any way I get access to` _viewSize`, and `_layout` (and maybe calcMatrixFor but maybe I should be doing that my self) in that function?

**Comment by JElgar** [comment]
Ah ok sorry, everything was null in the PdfViewerController because I wasn't constructing it in a stateful widget. Ive updated that and its working a treat with

```
            normalizeMatrix: (newValue) {
              const minScale = 1.0;
              const minFractionOnScreen = 0.2;

              final position = newValue.calcPosition(pdfController.viewSize);

              final newZoom = max(newValue.zoom, minScale);

              final hw = pdfController.viewSize.width / 2 / newZoom;
              final hh = pdfController.viewSize.height / 2 / newZoom;

              final minX =
                  -hw + pdfController.layout.documentSize.width * minFractionOnScreen;
              final maxX =
                  hw + pdfController.documentSize.width * (1 - minFractionOnScreen);

              final minY =
                  -hh + pdfController.layout.documentSize.height * minFractionOnScreen;
              final maxY =
                  hh + pdfController.layout.documentSize.height * (1 - minFractionOnScreen);

              final x = position.dx.clamp(minX, maxX);
              final y = position.dy.clamp(minY, maxY);

              return pdfController.calcMatrixFor(
                Offset(x, y),
                zoom: newZoom,
              );
            },
```

 Thanks so much!

**Comment by espresso3389** [comment]
Ah, do we need controller as a parameter? If so, I'll add it though it is a small breaking change.

**Comment by espresso3389** [comment]
1.0.87 adds a change to normalizeMatrix function to get more parameters:

```dart
PdfViewerParams(
 normalizeMatrix: (matrix, viewSize, layout, controller) {
    // If the controller is not ready, just return the input matrix.
    if (controller == null || !controller.isReady) return matrix;
    final position = newValue.calcPosition(viewSize);
    final newZoom = controller.params.boundaryMargin != null
      ? newValue.zoom
      : max(newValue.zoom, controller.minScale);
    final hw = viewSize.width / 2 / newZoom;
    final hh = viewSize.height / 2 / newZoom;
    final x = position.dx.range(hw, layout.documentSize.width - hw);
    final y = position.dy.range(hh, layout.documentSize.height - hh);
    return controller.calcMatrixFor(Offset(x, y), zoom: newZoom, viewSize: viewSize);
  },
),

**Comment by espresso3389** [comment]
As you know, you can manually wire controller to the function but these parameter may ease to implement the things without doing that.

**Comment by JElgar** [comment]
That sounds ideal but seems to actually break things. This code seems to work great:

With 1.0.86 this code works great but on 1.0.87 it seems to be broken.
```
            normalizeMatrix: (newValue) {
              // final position = newValue.calcPosition(viewSize);
              // return pdfController.calcMatrixFor(
              //   Offset(position.dx, position.dy),
              //   zoom: newValue.zoom,
              // );
              const minScale = 1.0;
              const minFractionOnScreen = 0.2;

              final position = newValue.calcPosition(pdfController.viewSize);

              final newZoom = max(newValue.zoom, minScale);

              final hw = pdfController.viewSize.width / 2 / newZoom;
              final hh = pdfController.viewSize.height / 2 / newZoom;

              final minX =
                  -hw + pdfController.layout.documentSize.width * minFractionOnScreen;
              final maxX =
                  hw + pdfController.documentSize.width * (1 - minFractionOnScreen);

              final minY =
                  -hh + pdfController.layout.documentSize.height * minFractionOnScreen;
              final maxY =
                  hh + pdfController.layout.documentSize.height * (1 - minFractionOnScreen);

              final x = position.dx.clamp(minX, maxX);
              final y = position.dy.clamp(minY, maxY);

              return pdfController.calcMatrixFor(
                Offset(x, y),
                zoom: newZoom,
              );
            },

```

I also tried on 1.0.87 using the parameters but that also don't work
```
            normalizeMatrix: (newValue, viewSize, layout, controller) {
              // final position = newValue.calcPosition(viewSize);
              // return pdfController.calcMatrixFor(
              //   Offset(position.dx, position.dy),
              //   zoom: newValue.zoom,
              // );
              const minScale = 1.0;
              const minFractionOnScreen = 0.2;

              final position = newValue.calcPosition(viewSize);

              final newZoom = max(newValue.zoom, minScale);

              final hw = viewSize.width / 2 / newZoom;
              final hh = viewSize.height / 2 / newZoom;

              final minX =
                  -hw + layout.documentSize.width * minFractionOnScreen;
              final maxX =
                  hw + layout.documentSize.width * (1 - minFractionOnScreen);

              final minY =
                  -hh + layout.documentSize.height * minFractionOnScreen;
              final maxY =
                  hh + layout.documentSize.height * (1 - minFractionOnScreen);

              final x = position.dx.clamp(minX, maxX);
              final y = position.dy.clamp(minY, maxY);

              return controller!.calcMatrixFor(
                Offset(x, y),
                zoom: newZoom,
              );
            },

```

By doesn't work I mean the limits seem to be as before (before 1.0.86) where there is no panning without zooming in

**Comment by JElgar** [comment]
Ah this seems to be it 😂 https://github.com/espresso3389/pdfrx/pull/251

**Comment by espresso3389** [comment]
Merged: #251 

And, 1.0.88 is just released.
